traffic_ops/traffic_ops_golang/parameter/parameters.go

```diff
@@ -0,0 +1,354 @@
+		return tc.DBError, tc.SystemError
+	}
+	resultRows, err := tx.NamedQuery(insertQuery(), pl)
+	if err != nil {
+		if pqErr, ok := err.(*pq.Error); ok {
+			err, eType := dbhelpers.ParsePQUniqueConstraintError(pqErr)
+			if eType == tc.DataConflictError {
+				return errors.New("a parameter with " + err.Error()), eType
+			}
+			return err, eType
+		}
+		log.Errorf("received non pq error: %++v from create execution", err)
+		return tc.DBError, tc.SystemError
+	}
+	defer resultRows.Close()
+
+	var id int
+	var lastUpdated tc.TimeNoMod
+	rowsAffected := 0
+	for resultRows.Next() {
+		rowsAffected++
+		if err := resultRows.Scan(&id, &lastUpdated); err != nil {
+			log.Error.Printf("could not scan id from insert: %s\n", err)
+			return tc.DBError, tc.SystemError
+		}
+	}
+	if rowsAffected == 0 {
+		err = errors.New("no parameter was inserted, no id was returned")
+		log.Errorln(err)
+		return tc.DBError, tc.SystemError
+	}
+	if rowsAffected > 1 {
+		err = errors.New("too many ids returned from parameter insert")
+		log.Errorln(err)
+		return tc.DBError, tc.SystemError
+	}
+
+	pl.SetID(id)
+	pl.LastUpdated = &lastUpdated
+	err = tx.Commit()
+	if err != nil {
+		log.Errorln("Could not commit transaction: ", err)
+		return tc.DBError, tc.SystemError
+	}
+	rollbackTransaction = false
+	return nil, tc.NoError
+}
+
+func insertQuery() string {
+	query := `INSERT INTO parameter (
+name,
+config_file,
+value,
+secure) VALUES (
+:name,
+:config_file,
+:value,
+:secure) RETURNING id,last_updated`
+	return query
+}
+
+func (parameter *TOParameter) Read(db *sqlx.DB, parameters map[string]string, user auth.CurrentUser) ([]interface{}, []error, tc.ApiErrorType) {
+	var rows *sqlx.Rows
+
+	// Query Parameters to Database Query column mappings
+	// see the fields mapped in the SQL query
+	queryParamsToQueryCols := map[string]dbhelpers.WhereColumnInfo{
+		"config_file":  dbhelpers.WhereColumnInfo{"p.config_file", nil},
+		"id":           dbhelpers.WhereColumnInfo{"p.id", api.IsInt},
+		"last_updated": dbhelpers.WhereColumnInfo{"p.last_updated", nil},
+		"name":         dbhelpers.WhereColumnInfo{"p.name", nil},
+		"secure":       dbhelpers.WhereColumnInfo{"p.secure", api.IsBool},
+	}
+
+	where, orderBy, queryValues, errs := dbhelpers.BuildWhereAndOrderBy(parameters, queryParamsToQueryCols)
+	if len(errs) > 0 {
+		return nil, errs, tc.DataConflictError
+	}
+
+	query := selectQuery() + where + ParametersGroupBy() + orderBy
+	log.Debugln("Query is ", query)
+
+	rows, err := db.NamedQuery(query, queryValues)
+	if err != nil {
+		log.Errorf("Error querying Parameters: %v", err)
+		return nil, []error{tc.DBError}, tc.SystemError
+	}
+	defer rows.Close()
+
+	params := []interface{}{}
+	for rows.Next() {
+		var s tc.ParameterNullable
+		if err = rows.StructScan(&s); err != nil {
+			log.Errorf("error parsing Parameter rows: %v", err)
+			return nil, []error{tc.DBError}, tc.SystemError
+		}
+		params = append(params, s)
+	}
+
+	return params, []error{}, tc.NoError
```

Comment:
Why return an empty array here instead of nil?

Reply:
for json decoding -- see above